jobrunner/run.py

```diff
@@ -195,28 +155,142 @@ def handle_running_job(job):
+        # final state - we have we have finished!
+        results = api.get_results(definition)
+        # TODO: implement workspace API
+        # delete_obsolete_files(job, results)
+        save_results(job, results)
+        mark_job_as_completed(job)
+        api.cleanup(definition)
+        # we are done
+        return
+
+    # following logic is common to all non final transitions
+
+    if new_status.state == old_status.state:
+        # no change in state, i.e. back pressure
+        set_message(job, new_status.message, code=StatusCode.WAITING_ON_WORKERS)
 
-        if is_running:
-            set_message(job, "Running")
+    elif new_status.state == next_state:
+        # successful state change to the expected next state
+        message = new_status.state.value.title()
+        if old_status == ExecutorState.PENDING:
+            # we have started!
+            mark_job_as_running(job, message)
         else:
-            mark_job_as_completed(job)
-            api.cleanup(job_to_job_definition(job))
-    except JobError as exception:
-        set_message(job, "Failed")
-        mark_job_as_failed(job, exception)
-        api.cleanup(job_to_job_definition(job))
-    except Exception:
-        set_message(job, "Failed")
-        mark_job_as_failed(job, "Internal error when finalising job")
-        # We don't clean up here, to facilitate with debugging this unexpected error
-        raise
+            set_message(job, message)
+
+    elif new_status.state == ExecutorState.ERROR:
+        # all transitions can go straight to error
+        mark_job_as_failed(job, new_status.message)
+        api.cleanup(definition)
+
+    else:
+        raise Exception(
+            f"unexpected state transition from {old_status.state} to {new_status.state}"
+        )
+
+
+def save_results(job, results):
+    """Example the results of the execution and update the job accordingly."""
+
+    # this logic is adapted directly from jobrunner.manage_jobs.finalize_job()
+    job.outputs = results.outputs
+
+    # Set the final state of the job
+    if results.exit_code != 0:
+        job.state = State.FAILED
+        job.status_message = "Job exited with an error code"
+        job.status_code = StatusCode.NONZERO_EXIT
+    elif results.unmatched_patterns:
+        job.state = State.FAILED
+        job.status_message = "No outputs found matching patterns:\n - {}".format(
+            "\n - ".join(results.unmatched_patterns)
+        )
+        # If the job fails because an output was missing its very useful to
+        # show the user what files were created as often the issue is just a
+        # typo
+
+        # TODO:  job.unmatched_outputs = ???
+    else:
+        job.state = State.SUCCEEDED
+        job.status_message = "Completed successfully"
+
+    # two updates is messy, but will do for now
+    update(job, update_fields=["state", "outputs", "unmatched_outputs"])
+    set_message(job, results.status_message, results.status_code)
```

Comment:
so, set_message contains non-trival timestamp update logic that I wanted to preserve. But perhas we do not need it in this specific case.

Reply:
Yeah, I don't think we do need it here -- isn't it just to avoid spamming updates while we're polling for something?

Reply:
I think knowing that your incoming branch changes update to update all the things, then I think I can just drop the update() call here? And let set_message do it?

Reply:
Attempted fix in f585f91

Reply:
I'm sorry to keep going on about this.

We're not just setting the message here -- we're storing all the results of running the job, including the list of outputs. It seems obscure and maybe fragile to rely on a `set_message()` function to have the side effect of persisting all the changes to the job. I really think it would be clearer to call `update()`.

`set_message()` does two other things. It updates `job.updated_at`, which we probably should do here and make sure we're doing elsewhere, and it has spam-avoidance logic which we don't need here.

Reply:
Ok, I was just trying to reuse the existing log as much as possible, but will change.